superset/commands/database/create.py

```diff
@@ -76,34 +77,22 @@ def run(self) -> Model:
             "{}",
         )
 
-        try:
-            database = DatabaseDAO.create(attributes=self._properties, commit=False)
-            database.set_sqlalchemy_uri(database.sqlalchemy_uri)
+        ssh_tunnel = None
 
-            ssh_tunnel = None
+        try:
             if ssh_tunnel_properties := self._properties.get("ssh_tunnel"):
                 if not is_feature_enabled("SSH_TUNNELING"):
-                    db.session.rollback()
                     raise SSHTunnelingNotEnabledError()
-                try:
-                    # So database.id is not None
-                    db.session.flush()
-                    ssh_tunnel = CreateSSHTunnelCommand(
-                        database.id, ssh_tunnel_properties
-                    ).run()
-                except (SSHTunnelInvalidError, SSHTunnelCreateFailedError) as ex:
-                    event_logger.log_with_context(
-                        action=f"db_creation_failed.{ex.__class__.__name__}.ssh_tunnel",
-                        engine=self._properties.get("sqlalchemy_uri", "").split(":")[0],
-                    )
-                    # So we can show the original message
-                    raise ex
-                except Exception as ex:
-                    event_logger.log_with_context(
-                        action=f"db_creation_failed.{ex.__class__.__name__}.ssh_tunnel",
-                        engine=self._properties.get("sqlalchemy_uri", "").split(":")[0],
-                    )
-                    raise DatabaseCreateFailedError() from ex
+
+                database = self._do_create_database()
```

Comment:
@geido why wouldn't we want to add the database to the session here?

Reply:
@eschutho the rollback will look for the last item added to the session. I want the database to be last, so if the creation of the ssh tunnel fails, the database will be rolled back. There are other ways to define what to rollback, using checkpoints for instance, however from my experience those will fire flush events, which will eventually cause the premature fire of the event listeners.

Reply:
Wouldn't it roll back everything in the session? Including both the db and ssh_tunnel if either of them raised?

Reply:
I could not get it to behave like that unfortunately. Let me see if I've missed anything